metrics/opentelemetry/metrics.go

```diff
@@ -0,0 +1,158 @@
+package opentelemetry
+
+import (
+	"errors"
+	"fmt"
+	"log/slog"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/gobeam/stringy"
+	"go.opentelemetry.io/otel/metric"
+)
+
+type (
+	// Elem represents a metric element.
+	Elem struct {
+		Name  string
+		Value reflect.Value
+		Tag   reflect.StructTag
+	}
+)
+
+// FindAllMetrics finds all metrics in the given struct.
+func FindAllMetrics(m any, prefix string) <-chan Elem {
+	ret := make(chan Elem, 16)
+
+	go func() {
+		defer close(ret)
+
+		v := reflect.Indirect(reflect.ValueOf(m))
+		q := []Elem{{Value: v, Name: prefix}}
+		for len(q) > 0 {
+			e := q[0]
+			q = q[1:]
+			if e.Value.Kind() == reflect.Struct {
+				t := e.Value.Type()
+				for i := 0; i < t.NumField(); i++ {
+					f := t.Field(i)
+					if f.Type.Kind() == reflect.Struct {
+						q = append(q, Elem{Name: e.Name + f.Name + ".", Value: e.Value.Field(i)})
+					} else {
+						ret <- Elem{Name: e.Name + f.Name, Value: e.Value.Field(i), Tag: f.Tag}
+					}
+				}
+			}
+		}
+	}()
+
+	return ret
+}
+
+// CreateMetric creates a metric.
+func CreateMetric[F func(string, ...O) (M, error), O, M any](name string, f F, opts ...metric.InstrumentOption) (M, error) {
+	fOpts := make([]O, len(opts))
+	for i, o := range opts {
+		fOpts[i] = o.(O)
+	}
+	return f(name, fOpts...)
+}
+
+// CreateHistogramMetric creates a histogram metric.
+func CreateHistogramMetric[F func(string, ...O) (M, error), O, M any](name string, f F, histOpts []metric.HistogramOption, opts ...metric.InstrumentOption) (M, error) {
+	histOptsLen := len(histOpts)
+	fOpts := make([]O, histOptsLen+len(opts))
+	for i, o := range histOpts {
+		fOpts[i] = o.(O)
+	}
+	for i, o := range opts {
+		fOpts[histOptsLen+i] = o.(O)
+	}
+	return f(name, fOpts...)
+}
+
+// MetricName returns the metric name in snake case.
+func MetricName(name string) string {
+	return stringy.New(name).SnakeCase(".", ".").ToLower()
+}
+
+// HistogramOptions returns the options for a histogram metric.
+func HistogramOptions(elem Elem) []metric.HistogramOption {
+	if bb, ok := elem.Tag.Lookup("buckets"); ok {
+		floatStrings := strings.Split(bb, ",")
+		floats := make([]float64, len(floatStrings))
+		var err error
+		for i, s := range floatStrings {
+			if floats[i], err = strconv.ParseFloat(s, 64); err != nil {
+				slog.Error("failed to parse bucket boundary", "boundary", s, "err", err.Error())
+				return nil
+			}
+		}
+		return []metric.HistogramOption{metric.WithExplicitBucketBoundaries(floats...)}
+	}
+	return nil
+}
+
+// InitMetrics initializes all metrics in the given struct (except those tagged with metric="-").
+func InitMetrics(meter metric.Meter, m any, prefix string) error {
```

Comment:
Curious, is this a required step or a pre-optimization similar to what we have for Prometheus?

Reply:
Similar to what you have for prometheus, in this case, it takes the nuisance about calling the initializer and checking errors for every bleeping metric by using reflection to do the necessary and choosing the name from its member name and hierarchy.